### PR TITLE
Added setting on admin to skip user verification

### PR DESCRIPTION
**Pull request by raul-fuentes**
Where
=====
* **Related Issue:** https://github.com/consul/consul/issues/2291

What
====
- It allows to ignore the user verification (postal mail, sms, telephone, email and census, all appear verified) using a new setting. If activated, any user will be considered verified and see the welcome page only in his first login. 

- Using seeds on a new installation sets this setting active (on seeds and dev_seeds)

How
===
- Added a new option to settings (skip_validation) which makes to returns true any methods from models/concerns/validation.rb wich involves any kind of validation

- Also added conditions to force the user to go to welcome_path only in his first login

Screenshots
===========
- As you can see after the activation of the new setting a unverified user can login, with all participation options active and see the welcome page only once

peek 2018-04-13 11-13

Test
====
- Added test to verify the correct activation of the setting
- Added test for any verification method skipped
- Added test to ensure that welcome view is showed one time to any user on his first login if the setting is active and shows correctly the four green checks about active users

Deployment
==========
- No

Warnings
========
- No

### CODE REVIEW

File: app/controllers/users/sessions_controller.rb [hunk 1]
```diff
@@ -3,13 +3,17 @@ class Users::SessionsController < Devise::SessionsController
   private
 
     def after_sign_in_path_for(resource)
-      if !verifying_via_email? && resource.show_welcome_screen?
+      if !verifying_via_email? && resource.show_welcome_screen? && !override?
```
Comment (voodoorai2000) [inline app/controllers/users/sessions_controller.rb]: We don't want to skip the welcome page, 
it's nice to show it to all users the first time they login

What we want is to display the green check for all actions
Have a look at the gif "Initial welcome page when you sign up" [here](https://github.com/consul/consul/issues/2291)

File: app/controllers/users/sessions_controller.rb [hunk 2]
```diff
@@ -3,13 +3,17 @@ class Users::SessionsController < Devise::SessionsController
...
         welcome_path
       else
         super
       end
     end
 
+    def override?
+      Setting["feature.user.override_verification"] == 'active'
```
Comment (voodoorai2000) [inline app/controllers/users/sessions_controller.rb]: How about something like this for the name of the setting?
`Setting["feature.user.skip_verification"] == true`

The true is also important, all other features are either true or nil

File: db/seeds.rb [hunk 3]
```diff
@@ -122,3 +122,5 @@
 
 # Related content
 Setting['related_content_score_threshold'] = -0.3
+
+Setting["feature.user.override_verification"] = nil
```
Comment (voodoorai2000) [inline db/seeds.rb]: As a default, better to set skip verification to `true` 
People need time to configure their verification system, 
and it's a common request to skip verification in the initial setup

File: config/locales/en/settings.yml [hunk 4]
```diff
@@ -42,6 +42,7 @@ en:
       legislation: Legislation
       user:
         recommendations: Recommendeds
+        override_verification: Override user verification
```
Comment (voodoorai2000) [inline config/locales/en/settings.yml]: This translation should be updated accordingly to the feature name

File: spec/support/verifiable.rb [hunk 5]
```diff
@@ -160,4 +160,66 @@
...
+      user.update(letter_requested_at: Time.current, letter_verification_code: nil)
+      expect(user.verification_letter_sent?).to eq(true)
+
+      user.update(letter_requested_at: nil, letter_verification_code: nil)
+      expect(user.verification_letter_sent?).to eq(true)
+    end
+
+  end
+
 end
```
Comment (voodoorai2000) [inline spec/support/verifiable.rb]: Great!

It be nice to also add an integration spec in [admin settings](https://github.com/consul/consul/blob/master/spec/features/admin/settings_spec.rb) for this issue

Something on the lines of an admin updating this setting 
and making sure a standard user can then perform an action that requires verification

File: app/models/concerns/verification.rb [hunk 6]
```diff
@@ -10,32 +10,37 @@ module Verification
     scope :incomplete_verification, -> { where("(users.residence_verified_at IS NULL AND users.failed_census_calls_count > ?) OR (users.residence_verified_at IS NOT NULL AND (users.unconfirmed_phone IS NULL OR users.confirmed_phone IS NULL))", 0) }
   end
 
+  def override(value)
```
Comment (voodoorai2000) [inline app/models/concerns/verification.rb]: The way we usually skip the normal behaviour is with a [Guard Clause](http://www.thechrisoshow.com/2009/02/16/using-guard-clauses-in-your-ruby-code/)
How about something like this?:
```
def skip_verification? 
  Setting["feature.user.skip_verification"].present?
end
```
```
def verification_email_sent?
  return true if skip_verification?
  email_verification_token.present?
end
```

**Comment by clairezed** [comment]
This is a feature we need in a project I'm involved in. 
Is there anything I could do/improve on this PR to help it beeing merged ?

**Comment by raul-fuentes** [comment]
Hi @clairezed, currently I'm finishing a few changes requested by @voodoorai2000 and when he approves my changes, this PR will be done :)

**Comment by raul-fuentes** [comment]
All fixed!

**Comment by bertocq** [comment]
Thanks for expressing your interest and offering help @clairezed ! This PR will be reviewed again soon and you can count it will be merged before next v0.15 release is done

File: app/controllers/users/sessions_controller.rb [hunk 7]
```diff
@@ -3,13 +3,17 @@ class Users::SessionsController < Devise::SessionsController
   private
 
     def after_sign_in_path_for(resource)
-      if !verifying_via_email? && resource.show_welcome_screen?
+      if !verifying_via_email? && (resource.show_welcome_screen? || (skip_verification? && resource.first_login?))
```
Comment (bertocq) [inline app/controllers/users/sessions_controller.rb]: Why are we adding logic here that should be included inside `show_welcome_screen?` ?

File: app/controllers/users/sessions_controller.rb [hunk 8]
```diff
@@ -3,13 +3,17 @@ class Users::SessionsController < Devise::SessionsController
...
         welcome_path
       else
         super
       end
     end
 
+    def skip_verification?
+      Setting["feature.user.skip_verification"]
```
Comment (bertocq) [inline app/controllers/users/sessions_controller.rb]: THere's a trailing `.present?` missing here, although this method won't be no longer needed here (and its a dup of the one at the verification concern

File: app/models/user.rb [hunk 9]
```diff
@@ -328,6 +328,10 @@ def interests
     follows.map{|follow| follow.followable.tags.map(&:name)}.flatten.compact.uniq
   end
 
+  def first_login?
```
Comment (bertocq) [inline app/models/user.rb]: There's already similar logic at  `show_welcome_screen?`, couldn't we just prevent duplicating code?

File: spec/support/verifiable.rb [hunk 10]
```diff
@@ -179,4 +179,66 @@
     end
   end
 
+  describe "#overrided_methods" do
+
+    let(:user) {create(:user)}
+
+    before do
+      Setting["feature.user.skip_verification"] = 'active'
```
Comment (bertocq) [inline spec/support/verifiable.rb]: We're using `'true'` values all along, and here `'active'` why?

File: spec/support/verifiable.rb [hunk 11]
```diff
@@ -179,4 +179,66 @@
     end
   end
 
+  describe "#overrided_methods" do
```
Comment (bertocq) [inline spec/support/verifiable.rb]: This is not consistent with rspec standards or the codebase.. Also "override" is not the verb agreed upon, but `skip`.

File: spec/support/verifiable.rb [hunk 12]
```diff
@@ -179,4 +179,66 @@
...
+
+      user.update(confirmed_phone: "123456789", residence_verified_at: false)
+      expect(user.level_two_verified?).to eq(true)
+    end
+
+    it "level_three_verified? is true  if overrided" do
+      expect(user.level_three_verified?).to eq(true)
+    end
+
+    it "unverified? is true if overrided" do
```
Comment (bertocq) [inline spec/support/verifiable.rb]: This subject is not consistent with the expectation! It says it should be `true` but the expectation wants a `false`... also `unverified?` method is NOT affected but the skip verification :/ Please read carefully code changes before commiting, and after opening the PR. Reviewers time is really important having to read tons of PR's, you only have to read one at a time before moving into the next issue.

File: spec/features/welcome_spec.rb [hunk 13]
```diff
@@ -68,4 +68,25 @@
     expect(page).to have_current_path(root_path)
   end
 
+  scenario 'is not shown to a regular user if the verification its overrided and the user has already logged in 1 time' do
+    user = create(:user, sign_in_count: 1)
+
+    Setting["feature.user.skip_verification"] = 'true'
```
Comment (bertocq) [inline spec/features/welcome_spec.rb]: Remember every time a spec alters a setting, it should revert the change back afterwards, flakys on random spec order are born because of this

File: spec/features/welcome_spec.rb [hunk 14]
```diff
@@ -68,4 +68,25 @@
...
+
+    login_through_form_as(user)
+
+    expect(page).to have_current_path(root_path)
+  end
+
+  scenario 'is showed to a regular user if the verification its overrided and the user logs in for the first time' do
```
Comment (bertocq) [inline spec/features/welcome_spec.rb]: Do we really need two scenarios for this? Just merge both in a

File: spec/features/admin/settings_spec.rb [hunk 15]
```diff
@@ -89,4 +89,48 @@
 
   end
 
-end
+  describe "Skip verification" do
+
+    scenario "deactivate verification", :js do
+      Setting["feature.user.skip_verification"] = 'true'
```
Comment (bertocq) [inline spec/features/admin/settings_spec.rb]: Again, setting and resetting a `Setting` value in a before & after blocks are the right way as you can see in allmost all  the codebase specs using this kind of behaviour. Please read similar specs to understand how to tackle a particular spec/code, you'll run into this problems again and again if not.

File: spec/features/admin/settings_spec.rb [hunk 16]
```diff
@@ -89,4 +89,48 @@
...
+
+      visit admin_settings_path
+
+      accept_alert do
+        find("#edit_setting_#{setting.id} .button").click
+      end
+
+      logout
+
+      user = create(:user, sign_in_count: 1)
```
Comment (bertocq) [inline spec/features/admin/settings_spec.rb]: Shouldn't both scenarios be the same? We're trying to see if the having the setting enabled/disabled makes a difference... but you're changing the conditions on both so ...what are we really checking here?

**Comment by raul-fuentes** [comment]
This is one of my first contributions to consul, give me a moment that I rewrite a few things with what I have been learning

File: spec/support/verifiable.rb [hunk 17]
```diff
@@ -179,4 +179,85 @@
...
+      end
+    end
+
+    describe "#level_three_verified?" do
+      it "is true if skipped" do
+        expect(user.level_three_verified?).to eq(true)
+      end
+    end
+
+    describe "#unverified?" do
```
Comment (bertocq) [inline spec/support/verifiable.rb]: Please take your time to read my comments again, `unverified?` has nothing to do neither its affected with the verification skip feature. Why are we testing that by default it returns false here?

File: spec/features/admin/settings_spec.rb [hunk 18]
```diff
@@ -89,4 +89,49 @@
...
+        find("#edit_setting_#{setting.id} .button").click
+      end
+
+      logout
+
+      user = create(:user)
+
+      expect(user.sms_verified?).to eq(false)
+
+      Setting["feature.user.skip_verification"] = nil
```
Comment (bertocq) [inline spec/features/admin/settings_spec.rb]: Both `Setting["feature.user.skip_verification"] = 'true'` can be moved onto a `before` block, and both `Setting["feature.user.skip_verification"] = nil` into an after block, as I said in the comments about this spec

**Comment by bertocq** [comment]
We've lost the changes at `app/controllers/users/sessions_controller.rb` and user model files, aren't those no longer neccesary?

File: spec/support/verifiable.rb (hunk 17 above)
Comment (raul-fuentes) [reply spec/support/verifiable.rb]: I'll remove that test, but that function it's affected by the skip setting.
Its defined 
```
def unverified?
  !level_two_or_three_verified?
end
```
and level_two_or_three_verified? like 
```
def level_two_or_three_verified?
  level_two_verified? || level_three_verified?
end
```
also affected by the skip setting.

I thought that I should test any function affected by my changes

**Comment by clairezed** [comment]
Hey there, it's me again. As far as I understand, there are still problems with this PR ? 

v0.15 hasn't been released yet, so this PR could still be part of it ? 

Once again, is there anything I could do to help ? Current state of this PR and what remains to be done is not really clear to me. My own need for this feature has become quite pressing, so really don't hesitate share your needs !

**Comment by voodoorai2000** [comment]
This is great @raul-fuentes
Thank you very much for the effort 💪 

@clairezed, @MarkEwert this PR is now merged into the main codebase
Feel free to update your code from the master branch 👍
